content/6_Advanced/Lagrange.mdx

```diff
@@ -27,6 +23,141 @@ adding lambda\*smth
+
+- A bidirectional edge from $i$ ($1 \le i \le N$) to $i+1$ with weight $A[i]$ and capacity $1$.
+
+$f(x)$ will be the maximum cost $x$-flow through the graph. We can repeatidly find the maximum cost augmenting path $x$ times to get our answer. Because the maximum cost as a function of flow is concave, $f(x)$ will be concave. You can read more about simulating cost flows [here](https://codeforces.com/blog/entry/118391).
+
+</Spoiler>
+
+Consider the following graphs of $f(x)$ and $f(x)-\lambda x$. In this example, we have $\lambda=5$.
+
+<iframe
+	src="https://www.desmos.com/calculator/ydynwc2fej?embed"
+	width="500px"
+	height="300px"
+	frameborder="0"
+/>
+
+Here is where the fact that $f(x)$ is concave comes in. Because the slope is non-increasing, we know that $f(x) - \lambda x$ will **first increase, then stay the same, and finally decrease**.
+
+Let $v(\lambda)$ be the optimal maximal achievable sum with $\lambda$ penalty and $c(\lambda)$ be the number of subarrays used to achieve $v(\lambda)$ (note that if there are multiple such possibilities, we set $c(\lambda)$ to be the **maximal** number of subarrays to achieve $v(\lambda)$). These values can be calculated in $\mathcal{O}(N)$ time using the dynamic programming approach described above.
+
+When we assign the penalty of $\lambda$, we are trying to find the maximal sum if creating a subarray reduces our sum by $\lambda$. So $v(\lambda)$ will be the **maximum** of $f(x) - \lambda x$ and $c(\lambda)$ will equal to the rightmost $x$ that **maximizes** $f(x) - \lambda x$.
+
+Given the shape of $f(x) - \lambda x$, we know that $f(x) - \lambda x$ will be maximized at all points where $\lambda$ is equal to the slope of $f(x)$ (these points are red in the graph above). If there are no such points it will be maximized at the rightmost point where the slope is less than $\lambda$. So this means that $c(\lambda)$ will be the rightmost $x$ at which the slope of $f(x)$ is still greater or equal to $\lambda$.
+
+
+Now we know exactly what $\lambda$ represents: $\lambda$ is the slope and $c(\lambda)$ is the rightmost $x$ at which the slope of $f(x)$ is still greater or equal to $\lambda$.
+
+We binary search for $\lambda$ and find the first $\lambda$ such that $c(\lambda) \ge K$. Let the optimal value be $\lambda_{\texttt{opt}}$. Then our answer is $v(\lambda_{\texttt{opt}}) + \lambda_{\texttt{opt}} K$. Note that this works even if $c(\lambda_{\texttt{opt}}) \neq K$ since  $c(\lambda_{\texttt{opt}})$ and $K$ will be on the same line with slope $\lambda_{\texttt{opt}}$ in that case.
+
+Because calculating $v(\lambda)$ and $c(\lambda)$ with the dynamic programming solution described above will take $\mathcal{O}(N)$ time, this solution runs in $\mathcal{O}(N\log{\sum A[i]})$ time.
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+#define ll long long
+
+const ll INF = (ll)300000 * 1000000000;
+
+int main() {
+	int n, k;
+	cin >> n >> k;
+
+	int a[n + 1];
```

Comment:
do you ever touch element 0 of this array?

Reply:
uhh no. i one indexed to make the dp transitions cleaner as i don't have to worry about i-1 being sus

Reply:
> worry about i-1 being sus

if the array was 0-indexed, should your dp access -1 in the first place?
right now if your dp is doing so, currently this code is silently erroring

Reply:
> right now if your dp is doing so, currently this code is silently erroring

wait my for loop starts at 1 so im not accessing -1. like basically everything is one indexed in the code. surely its good :trust:

> if the array was 0-indexed, should your dp access -1 in the first place?

if the array is 0 indexed i would be accessing -1 with the current code (since i would have to start at i=0) but ig it can be fixed by initializing the base case using A[0] but that seems kinda strange and less straight forward.

Reply:
idk i feel like making initializing the base case explicitly would be kinda nice

Reply:
wait like make everything zero indexed and initialize the first entry with A[0]:
dp[0][0] = {0, 0};
dp[0][1] = {A[0] - lambda, 1};

instead of what I'm currently doing now:
dp[0][0] = {0, 0};
dp[0][1] = {-INF, 0};

Reply:
yeah i don't see anything wrong with that